File: backend/app/schemas/avaliacoes.py
from datetime import datetime
from decimal import Decimal
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from app.models.enums import StatusAvaliacao, StatusSubmissao, TipoQuestao
from app.schemas.documento import Documento


class MateriaResponse(BaseModel):
    id: UUID
    nome: str
    slug: str | None = None
    cor_token_ui: str | None = None
    ordem: int | None = None


class MateriaCreate(BaseModel):
    nome: str = Field(min_length=1)
    slug: str | None = None
    cor_token_ui: str | None = None
    ordem: int | None = None


class MateriaPatch(BaseModel):
    nome: str | None = None
    slug: str | None = None
    cor_token_ui: str | None = None
    ordem: int | None = None


class AssuntoResponse(BaseModel):
    id: UUID
    nome: str
    ordem: int


class AssuntoCreate(BaseModel):
    nome: str = Field(min_length=1)
    ordem: int = 0


class AssuntoPatch(BaseModel):
    nome: str | None = None
    ordem: int | None = None


class PastaResponse(BaseModel):
    id: UUID
    nome: str
    resumo_status_texto: str | None = None
    total_avaliacoes: int | None = None
    total_submissoes: int | None = None


class PastaPatch(BaseModel):
    nome: str | None = None


class PastaCreate(BaseModel):
    nome: str = Field(min_length=1)


class AvaliacaoListItem(BaseModel):
    id: UUID
    titulo: str
    status: StatusAvaliacao
    turma_id: UUID | None = None
    turma_nome: str | None = None
    prazo_utc: datetime | None = None
    publicado_em: datetime | None = None
    total_submissoes: int | None = None
    total_alunos_turma: int | None = None


class AvaliacaoDetail(BaseModel):
    id: UUID
    pasta_id: UUID
    titulo: str
    status: StatusAvaliacao
    turma_id: UUID | None = None
    turma_nome: str | None = None
    prazo_utc: datetime | None = None
    publicado_em: datetime | None = None
    encerrada_em: datetime | None = None
    payload_editor: dict[str, Any] | None = None
    instrucoes_gerais: Documento | dict[str, Any] | None = None
    versao: int
    questoes: list["QuestaoResponse"] = []


class AvaliacaoPublicar(BaseModel):
    turma_id: UUID


class AvaliacaoCreate(BaseModel):
    titulo: str = Field(min_length=1)
    prazo_utc: datetime | None = None
    turma_id: UUID | None = None


class AvaliacaoPatch(BaseModel):
    titulo: str | None = None
    prazo_utc: datetime | None = None
    payload_editor: dict[str, Any] | None = None
    instrucoes_gerais: Documento | None = None


class QuestaoUpsert(BaseModel):
    id: UUID | None = None
    tipo: TipoQuestao
    ordem: int
    enunciado: str | None = None
    conteudo: Documento | None = None
    alternativas: list[str] | None = None
    resposta_correta: int | None = None
    peso: Decimal = Decimal("1")

    @model_validator(mode="after")
    def exige_texto(self) -> "QuestaoUpsert":
        if not self.conteudo and not (self.enunciado and self.enunciado.strip()):
            raise ValueError("Informe enunciado ou conteudo")
        return self


class QuestaoResponse(BaseModel):
    id: UUID
    ordem: int
    tipo: TipoQuestao
    enunciado: str
    conteudo: dict[str, Any] | None = None
    alternativas: list[str] | None = None
    resposta_correta: int | None = None
    peso: Decimal


class QuestoesBulkReplace(BaseModel):
    questoes: list[QuestaoUpsert]


class QuestaoOrdemItem(BaseModel):
    id: UUID
    ordem: int


class QuestoesReorder(BaseModel):
    ordens: list[QuestaoOrdemItem]


class ArvoreAssunto(BaseModel):
    id: UUID
    nome: str
    ordem: int
    pastas: list["ArvorePasta"]


class ArvorePasta(BaseModel):
    id: UUID
    nome: str
    avaliacoes: list[AvaliacaoListItem]


class ArvoreMateria(BaseModel):
    id: UUID
    nome: str
    assuntos: list[ArvoreAssunto]


SituacaoAvaliacaoAluno = Literal["pendente", "em_andamento", "concluida"]


class AlunoAvaliacaoDisponivel(BaseModel):
    id: UUID
    titulo: str
    turma_id: UUID
    turma_nome: str
    prazo_utc: datetime | None = None
    status_avaliacao: StatusAvaliacao
    status_submissao: StatusSubmissao | None = None
    situacao: SituacaoAvaliacaoAluno
    nota_decimal: Decimal | None = None
    percentual_acerto: float | None = None


class AlunoAvaliacaoView(BaseModel):
    id: UUID
    titulo: str
    prazo_utc: datetime | None = None
    questoes: list["QuestaoAlunoView"]
    submissao_id: UUID | None = None
    status_submissao: StatusSubmissao | None = None
    situacao: SituacaoAvaliacaoAluno | None = None
    somente_leitura: bool = False
    exibir_gabarito: bool = False
    nota_decimal: Decimal | None = None
    percentual_acerto: float | None = None
    total_questoes: int = 0
    questoes_corretas: int = 0
    respostas: list["RespostaQuestaoInput"] = []


class AvaliacaoDuplicar(BaseModel):
    pasta_id: UUID | None = None
    titulo: str | None = None


class AvaliacaoReabrir(BaseModel):
    prazo_utc: datetime | None = None


class SubmissaoResumoProfessor(BaseModel):
    submissao_id: UUID | None = None
    aluno_id: UUID
    aluno_nome: str
    situacao: SituacaoAvaliacaoAluno
    status_submissao: StatusSubmissao | None = None
    nota_decimal: Decimal | None = None
    percentual_acerto: float | None = None
    enviada_em: datetime | None = None


class SubmissoesAvaliacaoProfessor(BaseModel):
    total_alunos: int
    total_concluidas: int
    total_pendentes: int
    alunos: list[SubmissaoResumoProfessor]


class QuestaoAlunoView(BaseModel):
    id: UUID
    ordem: int
    tipo: TipoQuestao
    enunciado: str
    conteudo: dict[str, Any] | None = None
    alternativas: list[str] | None = None
    indice_resposta_aluno: int | None = None
    indice_gabarito: int | None = None
    acertou: bool | None = None


class RespostaQuestaoInput(BaseModel):
    questao_id: UUID
    valor_texto: str | None = None
    indice_selecionado: int | None = None


class SubmissaoPatch(BaseModel):
    respostas: list[RespostaQuestaoInput]


class SubmissaoResponse(BaseModel):
    id: UUID
    avaliacao_id: UUID
    status: StatusSubmissao
    nota_decimal: Decimal | None = None
    percentual_acerto: float | None = None
    enviada_em: datetime | None = None


AvaliacaoDetail.model_rebuild()
ArvoreMateria.model_rebuild()

File: frontend/lib/api/dtos/avaliacoes.ts
import type { StatusAvaliacao, StatusSubmissao, TipoQuestao } from "./common"

export interface MateriaResponse {
  id: string
  nome: string
  slug?: string | null
  cor_token_ui?: string | null
  ordem?: number | null
}

export interface MateriaCreate {
  nome: string
  slug?: string | null
  cor_token_ui?: string | null
  ordem?: number | null
}

export interface MateriaPatch {
  nome?: string | null
  slug?: string | null
  cor_token_ui?: string | null
  ordem?: number | null
}

export interface AssuntoResponse {
  id: string
  nome: string
  ordem: number
}

export interface AssuntoCreate {
  nome: string
  ordem?: number
}

export interface AssuntoPatch {
  nome?: string | null
  ordem?: number | null
}

export interface PastaResponse {
  id: string
  nome: string
  resumo_status_texto?: string | null
  total_avaliacoes?: number | null
  total_submissoes?: number | null
}

export interface PastaCreate {
  nome: string
}

export interface PastaPatch {
  nome?: string | null
}

export interface AvaliacaoListItem {
  id: string
  titulo: string
  status: StatusAvaliacao
  turma_id?: string | null
  turma_nome?: string | null
  prazo_utc?: string | null
  publicado_em?: string | null
  total_submissoes?: number | null
  total_alunos_turma?: number | null
}

export interface AvaliacaoPublicar {
  turma_id: string
}

export interface QuestaoResponse {
  id: string
  ordem: number
  tipo: TipoQuestao
  enunciado: string
  conteudo?: Record<string, unknown> | null
  alternativas?: string[] | null
  resposta_correta?: number | null
  peso: string | number
}

export interface AvaliacaoDetail {
  id: string
  pasta_id: string
  titulo: string
  status: StatusAvaliacao
  turma_id?: string | null
  turma_nome?: string | null
  prazo_utc?: string | null
  publicado_em?: string | null
  encerrada_em?: string | null
  payload_editor?: Record<string, unknown> | null
  instrucoes_gerais?: Record<string, unknown> | null
  versao: number
  questoes: QuestaoResponse[]
}

export interface AvaliacaoCreate {
  titulo: string
  prazo_utc?: string | null
}

export interface AvaliacaoPatch {
  titulo?: string | null
  prazo_utc?: string | null
  payload_editor?: Record<string, unknown> | null
}

export interface QuestaoUpsert {
  id?: string | null
  tipo: TipoQuestao
  ordem: number
  enunciado?: string
  conteudo?: Record<string, unknown> | null
  alternativas?: string[] | null
  resposta_correta?: number | null
  peso?: string | number
}

export interface QuestoesBulkReplace {
  questoes: QuestaoUpsert[]
}

export interface ChatMensagem {
  id: string
  papel: "usuario" | "assistente"
  conteudo: string
  criado_em?: string | null
}

export interface IaGerarRequest {
  mensagem: string
  material_ids: string[]
  pasta_ids: string[]
}

/** Questão emitida pela IA durante o streaming (evento `questao`). */
export interface IaQuestaoGerada {
  tipo: TipoQuestao
  ordem: number
  enunciado: string
  alternativas?: string[] | null
  resposta_correta?: number | null
  peso?: number
}

export interface QuestaoOrdemItem {
  id: string
  ordem: number
}

export interface QuestoesReorder {
  ordens: QuestaoOrdemItem[]
}

export interface ArvorePasta {
  id: string
  nome: string
  avaliacoes: AvaliacaoListItem[]
  total_submissoes?: number | null
  resumo_status_texto?: string | null
}

export interface ArvoreAssunto {
  id: string
  nome: string
  ordem: number
  pastas: ArvorePasta[]
}

export interface ArvoreMateria {
  id: string
  nome: string
  assuntos: ArvoreAssunto[]
}

export type SituacaoAvaliacaoAluno = "pendente" | "em_andamento" | "concluida"

export interface AlunoAvaliacaoDisponivel {
  id: string
  titulo: string
  turma_id: string
  turma_nome: string
  prazo_utc?: string | null
  status_avaliacao: StatusAvaliacao
  status_submissao?: StatusSubmissao | null
  situacao: SituacaoAvaliacaoAluno
  nota_decimal?: string | number | null
  percentual_acerto?: number | null
}

export interface QuestaoAlunoView {
  id: string
  ordem: number
  tipo: TipoQuestao
  enunciado: string
  conteudo?: Record<string, unknown> | null
  alternativas?: string[] | null
  indice_resposta_aluno?: number | null
  indice_gabarito?: number | null
  acertou?: boolean | null
}

export interface AlunoAvaliacaoView {
  id: string
  titulo: string
  prazo_utc?: string | null
  questoes: QuestaoAlunoView[]
  submissao_id?: string | null
  status_submissao?: StatusSubmissao | null
  situacao?: SituacaoAvaliacaoAluno | null
  somente_leitura?: boolean
  exibir_gabarito?: boolean
  nota_decimal?: string | number | null
  percentual_acerto?: number | null
  total_questoes?: number
  questoes_corretas?: number
  respostas?: RespostaQuestaoInput[]
}

export interface AvaliacaoDuplicar {
  pasta_id?: string | null
  titulo?: string | null
}

export interface AvaliacaoReabrir {
  prazo_utc?: string | null
}

export interface SubmissaoResumoProfessor {
  submissao_id?: string | null
  aluno_id: string
  aluno_nome: string
  situacao: SituacaoAvaliacaoAluno
  status_submissao?: StatusSubmissao | null
  nota_decimal?: string | number | null
  percentual_acerto?: number | null
  enviada_em?: string | null
}

export interface SubmissoesAvaliacaoProfessor {
  total_alunos: number
  total_concluidas: number
  total_pendentes: number
  alunos: SubmissaoResumoProfessor[]
}

export interface RespostaQuestaoInput {
  questao_id: string
  valor_texto?: string | null
  indice_selecionado?: number | null
}

export interface SubmissaoPatch {
  respostas: RespostaQuestaoInput[]
}

export interface SubmissaoResponse {
  id: string
  avaliacao_id: string
  status: StatusSubmissao
  nota_decimal?: string | number | null
  percentual_acerto?: number | null
  enviada_em?: string | null
}
